src/ndb.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (C) 2016 Swift Navigation Inc.
+ * Contact: Roman Gezikov <rgezikov@exafore.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <string.h>
+#include "ndb.h"
+
+#include "ndb_p1_section.h"
+#include "ndb_p3_section.h"
+
+ndb_file_version_t ndb_file_version = 0x00000001;
+
+
+#ifdef NDB_COMMON_STORAGE
+ephemeris_t            ndb_ephemeris[PLATFORM_SIGNAL_COUNT];
+ndb_element_metadata_t ndb_ephemeris_md[PLATFORM_SIGNAL_COUNT];
+almanac_t              ndb_almanac[PLATFORM_SIGNAL_COUNT];
+ndb_element_metadata_t ndb_almanac_md[PLATFORM_SIGNAL_COUNT];
+
+#define NDB_EPHE_FILE_NAME   "ephemerides"
+#define NDB_ALMA_FILE_NAME   "almanacs"
+
+ndb_file_t ndb_ephe_file = {
+    .name = NDB_EPHE_FILE_NAME,
+    .fh = -1
+};
+ndb_file_t ndb_alma_file = {
+    .name = NDB_ALMA_FILE_NAME,
+    .fh = -1
+};
+#endif
+
+ndb_section_if_t *ndb_first_section = &p1_section_if;
+
+enum ndb_op_code ndb_init()
+{
+    enum ndb_op_code op_code;
+    op_code = ndb_core_init();
+    if(NDB_ERR_NONE != op_code) {
+        return op_code;
+    }
+
+    /*
+     * Platform specific code. Section P1 is always active.
+     */
+    ndb_register_section(&p3_section_if);
+
+    return NDB_ERR_NONE;
+}
+
+void ndb_register_section(ndb_section_if_t* section_if)
+{
+    ndb_section_if_t **p = &ndb_first_section;
+    while(NULL != *p){
+        *p = (*p)->next_section;
```

Comment:
I think you want `p = &(*p)->next_section`. I just fixed fixed the decoder https://github.com/swift-nav/piksi_firmware/commit/6d22acecf12b61284c399e0e0518bf7e7d5b47d7.